### PR TITLE
feat: refactor to allow for parsing of any single leaf or array of leafs

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
## Summary
- Refactor to allow for any single leaf or array of leafs to be parsed in typedAudienceConditions and Experiment.audienceConditions

The "why", or other context.
For consistency across SDKs

## Test plan
No additional tests were written.

## Issues
- 3755

### CODE REVIEW

File: core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java [hunk 1]
```diff
@@ -32,6 +32,75 @@
 import java.util.Map;
 
 public class ConditionUtils {
+
+    static public <T> Condition parseConditions(Class<T> clazz, Object object) throws InvalidAudienceCondition {
```
Comment (nchilada) [inline core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java]: Is it guaranteed that one of the other overloaded `parseConditions` methods is used if `object` is an instance of `List<Object>` or `org.json.JSONArray`?

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java [hunk 2]
```diff
@@ -306,9 +304,8 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
             String id = audienceObject.getString("id");
             String key = audienceObject.getString("name");
             Object conditionsObject = audienceObject.get("conditions");
-            JSONArray conditionJson = (JSONArray)conditionsObject;
 
-            Condition conditions = ConditionUtils.<UserAttribute>parseConditions(UserAttribute.class, conditionJson);
+            Condition conditions = ConditionUtils.<UserAttribute>parseConditions(UserAttribute.class, conditionsObject);
```
Comment (nchilada) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java]: Nice! I can't comment in `parseAudiences`, but don't we need to update it too? We want it to pass `conditionsObject` straight to `ConditionUtils.parseConditions` without converting to `JSONArray conditionJson`.

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java [hunk 3]
```diff
@@ -319,8 +317,7 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
             String id = (String)audienceObject.get("id");
             String key = (String)audienceObject.get("name");
             Object conditionObject = audienceObject.get("conditions");
-            JSONArray conditionJson = (JSONArray)conditionObject;
-            Condition conditions = ConditionUtils.<UserAttribute>parseConditions(UserAttribute.class, conditionJson);
+            Condition conditions = ConditionUtils.<UserAttribute>parseConditions(UserAttribute.class, conditionObject);
```
Comment (nchilada) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java]: Nice! I can't comment in `parseAudiences`, but don't we need to update it too? We want it to pass `conditionsObject` straight to `ConditionUtils.parseConditions` without converting to `JSONArray conditionJson`.

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java (hunk 2 above)
Comment (thomaszurkan-optimizely) [reply core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java]: No audience.conditions is legacy and should never change.  This change is for typedAudiences and Experiment.audienceConditions as mentioned in the ticket.

File: core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java (hunk 1 above)
Comment (thomaszurkan-optimizely) [reply core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java]: Yes. As, tested in the current parsing of datafiles.

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java [hunk 4]
```diff
@@ -156,13 +156,11 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
...
                 Object jsonCondition = experimentObject.get("audienceConditions");
-                if (jsonCondition instanceof JSONArray) {
-                    try {
-                        conditions = ConditionUtils.<AudienceIdCondition>parseConditions(AudienceIdCondition.class, (JSONArray)jsonCondition);
-                    } catch (Exception e) {
-                        // unable to parse conditions.
-                        Logger.getAnonymousLogger().log(Level.ALL, "problem parsing audience conditions", e);
-                    }
+                try {
+                    conditions = ConditionUtils.<AudienceIdCondition>parseConditions(AudienceIdCondition.class, jsonCondition);
```
Comment (thomaszurkan-optimizely) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java]: @nchilada this is the only other one that will expect any type.  audience.conditions as we mentioned in the ticket and offline will remain legacy.

File: core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java (hunk 1 above)
Comment (nchilada) [reply core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java]: Cool, thanks!

**Pull request merged**